Auto-locate files to load in dev server, fix static build, expand documentation

kadamwhite opened the pull request:
As I prepare for WCUS I compared this to my previous Webpack dev server integration experiments and noticed that it didn't properly load built files when the dev server wasn't running. `react-scripts` uses the `webpack-manifest-plugin` to generate an assets manifest while running the build task, which we can digest to load all generated CSS and JS. This is a naive approach but satisfies the general use-case where all exported assets are intended to be loaded.

I've expanded the existing override wrapper for the `start` task to use the same plugin in development mode, as it can be configured to output an assets file even when used with `webpack-dev-server`. This let me expand the PHP in this repo to load the development server _or_ built files, depending on which manifest is available (and therefore whether the dev server or built code should be used).

For the time being I have removed the commented-out CSS code and attempted to clean up the PHP for consideration as a released package; it's early days but I would like to get this out there to help us (and other teams) use the project. It could use some good code review first, though :)

As I've worked through this I've updated the documentation with code comments, a conceptual walk-through in the README, and updated installation steps.

kadamwhite commented on loader.php:
```diff
@@ -15,45 +15,129 @@ function is_development() {
+ */
+function autoenqueue_plugin_assets() {
+	autoenqueue_assets( false );
+}
```
It would be great to have a universal way of getting the path or URI for a file regardless of whether it's in a theme or a plugin, but I am not aware of one. If there's a good option out there (or a better solution with what we have, if not) I'd welcome the input, I don't much like passing boolean flags all over the place :)

kadamwhite replied:
Had not seen #1, I can adapt that interface

kadamwhite commented on scripts/override-start.js:
```diff
@@ -1,83 +1,68 @@
+	// Permit dev server access from our WordPress host.
+	serverConfig.headers = Object.assign({}, serverConfig.headers, {
+		'Access-Control-Allow-Origin': '*',
+	});
```
In an ejected application or if we were not using `create-react-app`, [`allowedHosts`](https://webpack.js.org/configuration/dev-server/#devserver-allowedhosts) seems like it would be a better fit here; but this is the only one-size-fits-all solution.

rmccue replied:
We also don't know the WP host at the time of npm starting, so not sure we could do that anyway.

kadamwhite replied:
Exactly

kadamwhite commented on scripts/start.js:
```diff
@@ -3,22 +3,24 @@ const { spawn } = require( 'child_process' );
+	choosePort,
+	createCompiler,
+	prepareProxy,
+	prepareUrls,
```
This project exists in a middle-ground between the React project's style and WordPress's; I'd ask for _purely_ stylistic comments to be held until a later PR, I recognize there is inconsistency here :)

kadamwhite commented on loader.php:
```diff
@@ -15,45 +15,129 @@ function is_development() {
+ * @param bool $is_theme Whether to look for the manifest in a theme or plugin context.
+ * @return array|null;
+ */
+function get_assets_list( string $is_theme ) {
```
This function feels clunky but it's the only way I could see to achieve the desired fallback order of

1. If development, try to load and enqueue assets from dev server bundle
2. If not development or dev server assets bundle fails to load, try to load and enqueue assets from built bundle

rmccue commented on loader.php:
```diff
@@ -15,45 +15,129 @@ function is_development() {
+function get_file_path( string $path, bool $is_theme ) {
+	return $is_theme
+		? get_theme_file_path( $path )
+		: plugin_dir_path( __file__ ) . $path;
```
This should proper if/else, not a ternary.

rmccue commented on scripts/start.js:
```diff
@@ -27,30 +29,18 @@ fs.readFile( PORT_FILE, ( err, data ) => {
+		overrideConfigCache( `${ protocol }://${ HOST }:${ port }` );
+
+		// Pass the selected port forward so that react-scripts' start will not re-prompt.
+		process.env.PORT = port;
```
This should also set `process.env.BROWSER` to avoid launching the useless URL.

kadamwhite replied:
did so on line below, which didn't dismiss the notice :)

kadamwhite commented:
Good points all. I’ve seen a lot of webpack configs that yield a number of JS and css bundles; but it might be reasonable to count that as out-of-scope for the out of the box solution. It wouldn’t be too hard to ingest the manifest yourself if you have more complex needs; we could either just use one user specified slug for the two first css and js, or else we could process the bundle output to ensure the bundle name is passed through as the key and append that to the slug in a predictable way. (_e.g._ [userslug]_[bundlename]). Preference?

rmccue commented:
We're only really aiming at the create-react-app use case; once you've ejected, you're on your own.

I'd prefer we use an explicit slug passed via the PHP API, as noted in #1. Not a fan of the autoenqueue functions here either FWIW, explicit > implicit for a library like this.

kadamwhite commented:
@rmccue Ready for final review, I think:

- Adapted the API signature from #1, eliminated the automagic methods
- Removed the theme/plugin branching code in favor of using the $directory consistently
- Added the `BROWSER = 'none'` env suggestion
- Updated README accordingly

rmccue commented on README.md:
```diff
@@ -1,25 +1,89 @@
+- `scripts`: An array of script dependencies to load before your bundle.
+- `styles`: An array of stylesheet dependencies to load before your bundle.
+
+As in the examples above, for proper script resolution within a theme or plugin the `base_url` property is effectively required.
```
If it's required, can we make it a required parameter? I feel like it's actually possible to imply it from the path passed in though.

rmccue commented on loader.php:
```diff
@@ -15,45 +15,117 @@ function is_development() {
+ * @return array|null;
+ */
+function get_assets_list( string $directory ) {
+	$directory = trailingslashit( $directory );
```
I'd rather `untrailingslashit`, I generally prefer `$x . '/filename'` over `$x . 'filename'`, I think it reads a bit better.

kadamwhite replied:
The manifest file we use elsewhere contains paths without leading slashes, so it felt more appropriate to consistently have trailing-slash'd base strings and paths than to have it be situational. I'll change if you feel like it makes a difference but this feels more consistent to me..

rmccue commented on loader.php:
```diff
@@ -61,6 +61,29 @@ function( $asset_path ) { return 'build/' . $asset_path; },
+		return plugin_dir_url( __FILE__ ) . substr( $path, strlen( $plugin_path ) );
+	}
+
+	return '';
```
I'd generally prefer a `WP_Error` return (and maybe even `trigger_error`) for something like this, but #yolo for now.
